pypeit/datamodel.py

```diff
@@ -0,0 +1,911 @@
+"""
+Implements classes and function for the PypeIt data model.
+
+.. data-container:
+
+DataContainer
+-------------
+
+:class:`DataContainer` objects provide a utility for
+enforcing a specific datamodel on an object, and provides convenience
+routines for writing data to fits files. The class itself is an
+abstract base class that cannot be directly instantiated. As a base
+class, :class:`DataContainer` objects are versitile, but they have
+their limitations.
+
+Derived classes must do the following:
+
+    - Define a class attribute called ``datamodel``. See the examples
+      below for their format.
+    - Provide an :func:`__init__` method that defines the
+      instantiation calling sequence and passes the relevant
+      dictionary to this base-class instantiation.
+    - Provide a :func:`_validate` method, if necessary, that
+      processes the data provided in the `__init__` into a complete
+      instantiation of the object. This method and the
+      :func:`__init__` method are the *only* places where attributes
+      can be added to the class.
+    - Provide a :func:`_bundle` method that reorganizes the datamodel
+      into partitions that can be written to one or more fits
+      extensions. More details are provided in the description of
+      :func:`DataContainer._bundle`.
+    - Provide a :func:`_parse` method that parses information in one
+      or more fits extensions into the appropriate datamodel. More
+      details are provided in the description of
+      :func:`DataContainer._parse`.
+
+Here are some examples of how to and how not to use them.
+
+Basic container
++++++++++++++++
+
+Here's how to create a derived class for a basic container that
+holds two arrays and a metadata parameter::
+
+    import numpy as np
+    import inspect
+
+    from pypeit.datamodel import DataContainer
+
+    class BasicContainer(DataContainer):
```

Comment:
Can you make it clear what otype and atype are in these docs.

Reply:
Done.